src/modules/iotjs_module_tls.c

```diff
@@ -14,150 +14,583 @@
+      bio->read_idx = 0;
+      bio->write_idx = 0;
+      return 0;
+
+    default: { break; }
+  }
+
+  return SSL_BIO_ERROR;
+}
+
+
+int iotjs_bio_read(BIO *bio, const char *buf, size_t size) {
+  int sz;
+  char *pt;
+  sz = iotjs_bio_nread(bio, &pt, size);
+
+  if (sz > 0) {
+    memset((void *)buf, 0, (size_t)sz);
+    memcpy((void *)buf, pt, (size_t)sz);
+  }
+
+  return sz;
+}
+
+int iotjs_bio_write(BIO *bio, const char *buf, size_t size) {
+  int sz;
+  char *data;
+
+  sz = iotjs_bio_nwrite(bio, &data, size);
+
+  if (sz <= 0) {
+    return sz;
+  }
+
+  memset(data, 0, (size_t)sz);
+  memcpy(data, buf, (size_t)sz);
+  return sz;
+}
+
+
+BIO *iotjs_ssl_bio_new(int type) {
+  BIO *bio = (BIO *)iotjs_buffer_allocate(sizeof(BIO));
+  if (bio) {
+    bzero(bio, sizeof(BIO));
+    bio->type = type;
+    bio->mem = NULL;
+    bio->prev = NULL;
+    bio->next = NULL;
+  }
+  return bio;
+}
+
+int iotjs_bio_free(BIO *bio) {
+  if (bio) {
+    if (bio->pair != NULL) {
+      bio->pair->pair = NULL;
+    }
+    if (bio->mem)
+      iotjs_buffer_release((char *)bio->mem);
+    iotjs_buffer_release((char *)bio);
+  }
+  return 0;
+}
+
+int iotjs_bio_free_all(BIO *bio) {
+  while (bio) {
+    BIO *next = bio->next;
+    iotjs_bio_free(bio);
+    bio = next;
+  }
+  return 0;
+}
+
+int iotjs_bio_net_send(void *ctx, const unsigned char *buf, size_t len) {
+  BIO *bio = (BIO *)ctx;
+
+  int sz;
+  sz = iotjs_bio_write(bio, (const char *)buf, len);
+  if (sz <= 0) {
+    return MBEDTLS_ERR_SSL_WANT_WRITE;
+  }
+  return sz;
+}
+
+int iotjs_bio_net_recv(void *ctx, unsigned char *buf, size_t len) {
+  BIO *bio = (BIO *)ctx;
+  int sz;
+  sz = iotjs_bio_read(bio, (const char *)buf, len);
+
+  if (sz <= 0) {
+    return MBEDTLS_ERR_SSL_WANT_READ;
+  }
+  return sz;
+}
+
+JS_FUNCTION(TlsConstructor) {
+  DJS_CHECK_THIS();
 
   jerry_value_t jtls = JS_GET_THIS();
-  iotjs_tls_t* tls_data = tls_create(jtls);
+  iotjs_tls_t *tls_data = iotjs_tls_create(jtls);
+
+  jerry_value_t opts = jargv[0];
+
+  // Set initial handshake state
+  tls_data->handshake_state = SSL_HANDSHAKE_READY;
 
   int ret = 0;
+
   if ((ret = mbedtls_ctr_drbg_seed(&tls_data->ctr_drbg, mbedtls_entropy_func,
                                    &tls_data->entropy, NULL, 0)) != 0) {
-    return tls_connect_error(&port, &host, &hostname,
-                             "drbg seeding failed, error code: %d", ret);
+    return JS_CREATE_ERROR(COMMON, "drbg seeding failed");
   }
 
-  ret = mbedtls_net_connect(&tls_data->server_fd, iotjs_string_data(&host),
-                            iotjs_string_data(&port), MBEDTLS_NET_PROTO_TCP);
-  if (ret) {
-    return tls_connect_error(&port, &host, &hostname,
-                             "failed to connect to %s:%s, error code: %d",
-                             iotjs_string_data(&host), iotjs_string_data(&port),
-                             ret);
+  // User provided certificate
+  jerry_value_t jca_txt = iotjs_jval_get_property(opts, "ca");
```

Comment:
We're supposed to use the magic string for something like "ca" here. The same can be found for "rejectUnauthorized", "servername" and so on.